<a href="https://colab.research.google.com/github/UnaRam/pythonFundamemtals/blob/main/Una_Numpy_mini_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Investigate, wrangle, add new data to the dataframe, filter and correlate

The following data file contains data about student scores in math, reading and writing.  The exam scores are assumed to be percentages. It also contains data about gender, ethnicity, parental education, whether the student qualifies for free school food and whether or not the student has taken a preparation course for the exams.  The data set has already been cleaned. 

1.  Investigate the data set.

2.  Create numpy arrays to hold each of the three sets of scores.  
Create a new numpy array to hold the average exam score (of the three scores for each student)
Add the new numpy array as a new column in the dataframe.

3.  Filter the original dataset into a new dataframe containing just the females.  Calculate the average exam scores for all rows in this new dataframe and then find the mean of the average column.

4.  Do the same for the males.

5.  Use the original dataset to find the correlation coefficient for reading and maths.  How closely do they correlate?  Write what you find in a text box below the code.

6.  Do the same for reading and writing. What do you find?

7.  You might want to filter on different criteria and check correlation (e.g. those on free school meals, or those who had prepared, etc)







The dataset can be accessed here:  https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv.  This is a .csv file.

**NOTE:** Some useful references are included at the bottom of this spreadsheet.

Use the code cell below to work your code.

### Helpful references
---

Filtering on criteria using df references:  
`filtered_df = df[df['column name' == value]]`  

Filtering on multiple criteria using df reference:  
`filtered_df = df[df['first column name' == value] & df[second column name] >= value]`

Numpy help sheet:  http://datacamp-community-prod.s3.amazonaws.com/da466534-51fe-4c6d-b0cb-154f4782eb54



In [76]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

def calculate_average_scores(df):
  # print(df.info())
  # print(df.head())
  # print(df.tail())
  math_scores_a = np.array(df['math_score'])
  reading_scores_a = np.array(df['reading_score'])
  writing_scores_a = np.array(df['writing_score'])
  # print(math_scores_a.shape)
  # print(reading_scores_a.shape)
  # print(writing_scores_a.shape)
  average_score_a = (math_scores_a + reading_scores_a + writing_scores_a)/3
  # print(average_score_a)
  df['average_score']=average_score_a.tolist()

  # print("average score_a:\n",average_score_a)
  return average_score_a





def filter_df(df,col_to_filt,text_to_filter):
  # Filter the original dataset into a new dataframe containing just the females. 
  # Calculate the average exam scores for all rows in this new dataframe and then find the mean of the average column.
 
    filtered_df = df[df[col_to_filt] == text_to_filter]
    print("\n\nfiltered on ",text_to_filter,": \n", filtered_df.head())
    return filtered_df

#read raw data
url = "https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv"
scores_df = pd.read_csv(url)



def filter_and_return_average(df,col_to_filter,text_to_filter):
  filtered_scores_df = filter_df(df,col_to_filter,text_to_filter)
  filtered_average_a = calculate_average_scores(filtered_scores_df)
  filtered_average = round(filtered_average_a.mean(),2)
  return filtered_average

female_average = filter_and_return_average(scores_df,'gender','female')
print("FEMALE: ",female_average)
male_average = filter_and_return_average(scores_df,'gender','male')
print("MALE: ",male_average)
 



filtered on  female : 
    gender ethnicity parental_education         lunch preparation_course  math_score  reading_score  writing_score
0  female   group E       some college  free/reduced               none          65             76             71
2  female   group B   some high school  free/reduced          completed          62             56             61
4  female   group C        high school  free/reduced          completed          34             54             55
5  female   group B       some college  free/reduced               none          53             66             61
6  female   group D       some college      standard               none          85             88             92
FEMALE:  70.82


filtered on  male : 
    gender ethnicity  parental_education     lunch preparation_course  math_score  reading_score  writing_score
1    male   group C        some college  standard          completed          75             72             69
3    male   group D    some h

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [87]:
# 5. Use the original dataset to find the correlation coefficient for reading and maths. How closely do they correlate? Write what you find in a text box below the code.
def find_corr_coeff(df,col1,col2):
  # print(df.head())
  #split into arrays 
  x_array,y_array = split_into_arrays(df,col1,col2)
  # print("X:",x_array.shape)
  # print("Y:",y_array.shape)
  corr_coefficient = np.corrcoef(x_array,y_array)
  # print("corr:",corr_coefficient)
  return round(corr_coefficient[0][1],5)

def split_into_arrays(df,col1,col2):
  # print("col1: ",col1)
  # print("col2:",col2)
  first_a = np.array(df[col1])
  second_a = np.array(df[col2])
  return first_a, second_a


reading_maths_corr = find_corr_coeff(scores_df,'reading_score','math_score') 
print("\ncorrelation between reading and maths: ",reading_maths_corr)


correlation between reading and maths:  0.8146


correlation is +0.815 implying a fairly strong positive correlation between reading and maths

In [61]:
# 6. Do the same for reading and writing. What do you find?
reading_writing_corr = find_corr_coeff(scores_df,'reading_score','writing_score') 
print("\ncorrelation between reading and writing: ",reading_writing_corr)

   gender ethnicity parental_education         lunch preparation_course  math_score  reading_score  writing_score
0  female   group E       some college  free/reduced               none          65             76             71
1    male   group C       some college      standard          completed          75             72             69
2  female   group B   some high school  free/reduced          completed          62             56             61
3    male   group D   some high school      standard          completed          60             60             59
4  female   group C        high school  free/reduced          completed          34             54             55

correlation between reading and writing:  0.957


correlation coefficient between reading and writing is a very strong positive value of 0.957
This is stronger than the correlation between reading and maths

In [73]:
# You might want to filter on different criteria and check correlation (e.g. those on free school meals, or those who had prepared, etc)
#somewhere i have overwritten the scores_df so re-reading it in - but should go back and find out where
# url = "https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv"
# scores_df = pd.read_csv(url)
# print(scores_df.head())
filter_on_lunch_df= filter_df(scores_df,'lunch','free/reduced')
# print(filter_on_lunch_df.head(10))
reading_writing_corr_freelunch = find_corr_coeff(filter_on_lunch_df,'reading_score','writing_score') 
print("\ncorrelation between reading and writingfor filtered on lunch: ",reading_writing_corr_freelunch)

filtered on  free/reduced : 
     gender ethnicity  parental_education         lunch preparation_course  math_score  reading_score  writing_score
0   female   group E        some college  free/reduced               none          65             76             71
2   female   group B    some high school  free/reduced          completed          62             56             61
4   female   group C         high school  free/reduced          completed          34             54             55
5   female   group B        some college  free/reduced               none          53             66             61
11  female   group D  associate's degree  free/reduced               none          54             69             70
    gender ethnicity  parental_education         lunch preparation_course  math_score  reading_score  writing_score
0   female   group E        some college  free/reduced               none          65             76             71
2   female   group B    some high school  

again a high postive correlation between reading and writing when the students are filtered to those who get free/reduced lunch. 

In [88]:
filter_on_prepared_df= filter_df(scores_df,'preparation_course','completed')
# print("prep:\n",filter_on_prepared_df.info())
reading_writing_corr_prepared = find_corr_coeff(filter_on_prepared_df,'reading_score','writing_score') 
print("\ncorrelation between reading and writing for filtered on prepared: ",reading_writing_corr_prepared)

filter_on_not_prepared_df= filter_df(scores_df,'preparation_course','none')
# print("not prep:\n",filter_on_not_prepared_df.info())
reading_writing_corr_not_prepared = find_corr_coeff(filter_on_not_prepared_df,'reading_score','writing_score') 
print("\ncorrelation between reading and writing for filtered on not prepared: ",reading_writing_corr_not_prepared)



filtered on  completed : 
    gender ethnicity  parental_education         lunch preparation_course  math_score  reading_score  writing_score
1    male   group C        some college      standard          completed          75             72             69
2  female   group B    some high school  free/reduced          completed          62             56             61
3    male   group D    some high school      standard          completed          60             60             59
4  female   group C         high school  free/reduced          completed          34             54             55
8  female   group C  associate's degree      standard          completed          58             71             63

correlation between reading and writing for filtered on prepared:  0.9565


filtered on  none : 
     gender ethnicity parental_education         lunch preparation_course  math_score  reading_score  writing_score
0   female   group E       some college  free/reduced              

the correlation coefficients were very close so I changed the function to show 5 dp instead of 3 to troubleshoot - they are in fact different. Also used a print of the shape of the array to indicate number of items in the filtered list to make sure I was filtering properly
